pyrealm/pmodel/calc_optimal_chi_new.py

```diff
@@ -0,0 +1,573 @@
+
+    def summarize(self, dp: int = 2) -> None:
+        """Print CalcOptimalChi summary.
+
+        Prints a summary of the variables calculated within an instance
+        of CalcOptimalChi including the mean, range and number of nan values.
+
+        Args:
+            dp: The number of decimal places used in rounding summary stats.
+        """
+
+        attrs = [
+            ("xi", " Pa ^ (1/2)"),
+            ("chi", "-"),
+            ("mc", "-"),
+            ("mj", "-"),
+            ("mjoc", "-"),
+        ]
+        summarize_attrs(self, attrs, dp=dp)
+
+    @classmethod
+    def __init_subclass__(cls, method: str, is_c4: bool) -> None:
+        """Initialise a subclass deriving from this ABC."""
+
+        cls.method = method
+        cls.is_c4 = is_c4
+        OPTIMAL_CHI_CLASS_REGISTRY[cls.method] = cls
+
+
+# TODO - rootzonestress handling - new methods? Move rzs into env?
+
+
+class OptimalChiPrentice14(NewCalcOptimalChi, method="prentice14", is_c4=False):
+    r"""Calculate :math:`\chi` for C3 plants following :cite:`Prentice:2014bc`.
+
+    Optimal :math:`\chi` is calculated following Equation 8 in
+    :cite:`Prentice:2014bc`:
+
+    .. math:: :nowrap:
+
+        \[
+            \begin{align*}
+                \chi &= \Gamma^{*} / c_a + (1- \Gamma^{*} / c_a)
+                    \xi / (\xi + \sqrt D ), \text{where}\\
+                \xi &= \sqrt{(\beta (K+ \Gamma^{*}) / (1.6 \eta^{*}))}
+            \end{align*}
+        \]
+
+    The :math:`\ce{CO2}` limitation term of light use efficiency (:math:`m_j`) is
+    calculated following Equation 3 in :cite:`Wang:2017go`:
+
+    .. math::
+
+        m_j = \frac{c_a - \Gamma^{*}}
+                {c_a + 2 \Gamma^{*}}
+
+    Finally,  :math:`m_c` is calculated, following Equation 7 in
+    :cite:`Stocker:2020dh`, as:
+
+    .. math::
+
+        m_c = \frac{c_i - \Gamma^{*}}{c_i + K},
+
+    where :math:`K` is the Michaelis Menten coefficient of Rubisco-limited
+    assimilation.
+
+    Examples:
+        >>> import numpy as np
+        >>> env = PModelEnvironment(
+        ...     tc=np.array([20]), vpd=np.array([1000]),
+        ...     co2=np.array([400]), patm=np.array([101325.0])
+        ... )
+        >>> vals = OptimalChiPrentice14(env=env)
+        >>> vals.chi.round(5)
+        array([0.69435])
+        >>> vals.mc.round(5)
+        array([0.33408])
+        >>> vals.mj.round(5)
+        array([0.7123])
+        >>> vals.mjoc.round(5)
+        array([2.13211])
+    """
+
+    def set_beta(self) -> None:
+        """Set ``beta`` to a constant C3 specific value."""
+        # leaf-internal-to-ambient CO2 partial pressure (ci/ca) ratio
+        self.beta = self.pmodel_const.beta_cost_ratio_prentice14
+
+    def estimate_chi(self, xi_values: Optional[NDArray] = None) -> None:
+        """Estimate ``chi`` for C3 plants."""
+
+        if xi_values is not None:
+            _ = check_input_shapes(self.env.ca, xi_values)
```

Comment:
Maybe no need to assign the output to `_`.

Reply:
I tend to use that approach to make it clear that the function does return something, rather than altering its arguments. We could make `check_input_shape` which returns None and simply wraps `get_input_shapes` and drops the return value.